_posts/2025-10-23-mmaler-blogpost-2-quarkus-runtime-and-framework-for-cloud-native-java.adoc

```diff
@@ -0,0 +1,161 @@
+
+On top of that, Quarkus integrates established Java frameworks, such as CDI, JAX-RS, and Hibernate. However, it also enforces its own conventions, uses annotations to drive behavior, and defines how applications are structured.
+
+In this sense, Quarkus is a framework that serves as a runtime and platform for cloud-native Java applications.
+It also adheres to industry standards, offering compatibility with technologies such as Spring, Kafka, and Camel to support familiar and flexible development models.
+
+== Quarkus as a versatile framework
+
+Beyond staple traits of modern frameworks, Quarkus introduces two platform-defining features: build-time optimization and deep extensibility.
+
+* *Build-time optimization*: Quarkus shifts work from runtime to build time wherever possible.
+This approach reduces startup overhead and memory usage, resulting in a lean, fast, and efficient application tailored for production.
++
+image::2-build-time-principle.png[alt="Build-time principle in Quarkus: being fast by doing less at runtime", width=100%]
++
+image::3-BuildTimeP-Benefits-JVM.png[alt="Benefits of Quarkus build-time processing on the JVM", width=100%]
++
+image::4-BuildTimeP-Benefits-native.png[alt="Benefits of Quarkus native image compared to JVM", width=100%]
++
+image::5-JVM-Snapshotting-Native.png[alt="JVM snapshotting vs native images in Quarkus", width=100%]
+
+* *Extensibility*: Quarkus exposes extension points for everything from startup hooks to request filters.
+Over 800 extensions allow seamless integration with modern technologies such as Kafka, OpenTelemetry, and OpenID Connect.
++
+These extensions integrate with Quarkus and participate in its build-time and runtime lifecycle, making them first-class citizens of the Quarkus platform.
+
+=== Simplified developer experience
+
+Frameworks succeed when they reduce complexity without sacrificing power.
+Quarkus does exactly that:
+
+* Preconfigures popular libraries with sensible defaults.
+* Offers unified configuration and developer tooling.
+* Provides instant feedback with live reload and continuous testing.
+
+This makes Quarkus both powerful and approachable.
+
+You can start with a simple REST endpoint and scale it into a production-grade service without changing your development model.
+
+image::6-Frictionless.png[alt="Frictionless developer experience: live reload, dev services, continuous testing", width=100%]
+
+These features reflect Quarkus’s dual role: it behaves like a framework during development, offering structure, defaults, and conventions, and it acts as a runtime during execution, delivering performance, efficiency, and operational consistency.
```

Comment:
Not sure we can use this framework/runtime duality. Java frameworks have a slightly different meaning. I would use development framework, but that creates repetition.

Reply:
Would this replacement do?

These features give developers structure, sensible defaults, and clear conventions during development, and they deliver fast startup, low memory use, and operational consistency in production.

Or this one:
Together, these features give you a consistent workflow from code to production, with clear patterns while you build and predictable behavior when you run.